text/0225-configuration.md

```diff
@@ -0,0 +1,335 @@
+
+## Explanation
+
+Using a configuration model or configuration file, users can configure all options currently available via environment variables.
+
+### Goals
+
+* The configuration must be language implementation agnostic. It must not contain structure or statements that only can be interpreted in a subset of languages. This does not preclude the possibility that the configuration can have specific extensions included for a subset of languages, but it does mean that the configuration must be interpretable by all implementation languages.
+* Broadly supported format. Ideally, the information encoded in the file can be decoded using native tools for all OpenTelemetry implementation languages. However, it must be possible for languages that do not natively support an encoding format to write their own parsers.
+* The configuration format must support structured data. At the minimum arrays and associative arrays.
+* The format must support at least null, boolean, string, double precision floating point (IEEE 754-1985), and signed 64 bit integer value types.
+* Custom span processors, exporters, samplers, or other user defined extension components can be configured using this format.
+* Configure SDK, but also configure instrumentation.
+* Must offer stability guarantees while supporting evolution.
+* The structure of the configuration can be validated via a schema.
+* Support environment variable substitution to give users the option to avoid storing secrets in these files.
+
+#### Out of scope
+
+* Embedding additional configuration files within a configuration file through an expression. Additional configuration providers MAY choose to support this use-case in the future.
+
+## Internal details
+
+The schema for OpenTelemetry configuration is to be published in a repository to allow language implementations to leverage that definition to automatically generate code and/or validate end-user configuration. This will ensure that all implementations provide a consistent experience for any version of the schema they support. An example of such a proposed schema is available [here](https://github.com/MrAlias/otel-schema/tree/main/json_schema/schema).
+
+The working group proposes the use of [JSON Schema](https://json-schema.org/) as the language to define the schema. It provides:
+
+* support for client-side validation
+* code generation
+* broad support across languages
+
+In order to provide a minimal API surface area, implementations *MUST* support the following methods.
+
+### Configure(config)
+
+An API called `Configure` receives a configuration object. This method applies the configuration object's details to the SDK. This method specifically applies the configuration object to allow for multiple configuration format providers to be supported in the future. This OTEP describes two such providers in a file and data structure formats below, but remote file formats *MAY* be implemented in the future.
```

Comment:
```suggestion
### Configure(config) -> (TracerProvider, MeterProvider, LoggerProvider)

An API called Configure receives a configuration object and returns SDK components (TracerProvider, MeterProvider, LoggerProvider).

The resulting SDK component configuration SHOULD strictly reflect the configuration object's details and ignore the [opentelemetry environment variable configuration scheme](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/sdk-environment-variables.md).

This method can be invoked multiple times using different configuration arguments, with each invocation producing a separate set of SDK components. It is the caller's responsibility to ensure the resulting SDK components are correctly wired into the application and instrumentation. 
```

Reply:
I like this more than `global.Configure()`, which is by design required to have undocumented side effects. 

But it still does not address the concern of supporting dynamic configuration, because after the initial Configure() call the rest of the application would be already initialized with specific providers. Dynamic configuration means changing the _behavior_ of already instantiated objects, not creating new objects. I would rather have something like this:
```
interface Configurator {
    // apply can be called multiple times and may change the behavior of previously returned providers.
    void apply(Configuration);

    TracerProvider  getTracerProvider();
    MeterProvider   getTracerProvider();
    LoggerProvider getTracerProvider();
}
```

Reply:
So in the interface you describe, the getters initially return empty providers, and invocations of `apply` update the providers with the `Configuration`?

If so, this is quite similar to proposal I made [here](https://github.com/open-telemetry/oteps/pull/225#discussion_r1129622820) to add a `update(Configuration, TracerProvider, MeterProvider, LoggerProvider)` operation in the future, which when invoked, would apply the configuration to the providers passed as arguments. 

I feel like either way we spec it, an implementation would be compliant doing what you describe or what I describe. 

I'm happy to try to find words to describe the interface you propose, and use the code snippet you provide as an example.

Reply:
+1. To avoid the empty providers (it's weird indeed):
```
interface Configurator {
    TracerProvider  getTracerProvider();
    MeterProvider   getMeterProvider();
    LoggerProvider  getLoggerProvider();
}

cfgr = new StaticConfigurator(config)

class DynamicConfigurator extends Configurator {
   // apply can be called multiple times and may change the behavior
   // of previously returned providers.
   void update(Configuration);
}

cfgr = new DynamicConfigurator(config)
cfgr.update(config)
```